open-ce/conda_env_file_generator.py

```diff
@@ -0,0 +1,108 @@
+
+class CondaEnvFileGenerator():
+    """
+    The CondaEnvData class holds all of the information needed to generate a conda
+    environment file while can be used to create a conda environment from the built packages.
+    """
+
+    #pylint: disable=too-many-instance-attributes,too-many-arguments
+    def __init__(self,
+                 python_versions=utils.DEFAULT_PYTHON_VERS,
+                 build_types=utils.DEFAULT_BUILD_TYPES,
+                 channels=None,
+                 output_folder=None,
+                 env_file_prefix=utils.CONDA_ENV_FILENAME_PREFIX,
+                 ):
+        self.python_versions = python_versions
+        self.build_types = build_types
+        self.env_file_prefix = env_file_prefix
+        self.dependency_dict = {}
+        self.channels = []
+        self._initialize_dependency_dict()
+        self._initialize_channels(channels, output_folder)
+
+    def _initialize_dependency_dict(self):
+        for py_version in utils.parse_arg_list(self.python_versions):
+            for build_type in utils.parse_arg_list(self.build_types):
+                key = "py" + py_version + "-" + build_type
+                self.dependency_dict[key] = set()
+
+    def _initialize_channels(self, channels, output_folder):
+        self.channels.append("file:/" + output_folder)
+        if channels is None:
+            channels = []
+        for channel in channels:
+            self.channels.append(channel)
+        self.channels.append("defaults")
+
+    def write_conda_env_files(self, path=os.getcwd()):
+        """
+        This function writes conda environment files using the dependency dictionary
+        created from all the buildcommands.
+        """
+
+        conda_env_files = []
+        for key in self.dependency_dict:
+            if len(self.dependency_dict[key]) == 0:
+                continue
+            if not os.path.exists(path):
+                os.mkdir(path)
+            conda_env_name = self.env_file_prefix + key + ".yaml"
+            conda_env_file = os.path.join(path, conda_env_name)
+            data = dict(
+                name = conda_env_name,
+                channels = self.channels,
+                dependencies = self.dependency_dict[key],
+            )
+            with open(conda_env_file, 'w') as outfile:
+                yaml.dump(data, outfile, default_flow_style=False)
+                conda_env_files.append(conda_env_file)
+
+        return conda_env_files
+
+    def _update_deps_lists(self, dependencies, key):
+        if not dependencies is None:
+            for dep in dependencies:
+                self.dependency_dict[key].add(_cleanup_depstring(dep))
+
+    def update_conda_env_file_content(self, build_command, build_tree):
+        """
+        This function updates dependency dictionary for each build command with
+        its dependencies both internal and external.
+        """
+
+        key = "py" + build_command.python + "-" + build_command.build_type
```

Comment:
This key is getting generated in several places. It will get harder to maintain as we add more variants. Maybe it could be a method to build command (something like `build_command.variant_key()`). Then we would only have to update it in one place when more variants are added (see #90 )

Reply:
Yes, I'll create a method in BuildCommand for this. One thing I wanted to know is if BuildCommand will ever have python and build_type as None.

Reply:
I've instead created the method in utils.py as this variant key was needed irrespective of BuildCommand.